nvim: follow/unfollow participants

MichaelBitard opened the pull request:
This is a first raw implementation of the follow mode, works on my machine :tm: (I'll try it with a coworker monday probably).

#58 

I don't mind changing a lot of stuff (rename, code convention,...) if you want.

Anyway, I can't wait for my coworker to come back from holidays to try ethersync in nvim!!!! 

Awesome project!

## Questions

* I'm not really satisfied with the unfollow, I'd like to automatically unfollow when I type Esc, but I don't know how to do that, is it possible?
* In the readme, I updated the mappings examples but I think the keys I used are used by other plugins (I use Lazyvim and f and u are already mapped), which key combination should we set?

MichaelBitard commented:
The default change view is quite unreadable around the set_cursor method, don't forget to check "hide whitespace", I just removed the return, put the rest in an else in order to be able to follow in every condition

MichaelBitard commented:
For my future self, we got a bunch of errors:

```
Error executing vim.schedule lua callback: /usr/share/nvim/runtime/lua/vim/lsp/util.lua:947: Cursor position outside buffer
stack traceback:
	[C]: in function 'nvim_win_set_cursor'
	/usr/share/nvim/runtime/lua/vim/lsp/util.lua:947: in function 'show_document'
	...workspace/ethersync/nvim-plugin/lua/ethersync/cursor.lua:45: in function 'jump'
	...workspace/ethersync/nvim-plugin/lua/ethersync/cursor.lua:57: in function 'follow'
	...workspace/ethersync/nvim-plugin/lua/ethersync/cursor.lua:148: in function 'set_cursor'
	/opt/workspace/ethersync/nvim-plugin/plugin/ethersync.lua:60: in function 'process_operation_for_editor'
	/opt/workspace/ethersync/nvim-plugin/plugin/ethersync.lua:83: in function ''
	vim/_editor.lua: in function <vim/_editor.lua:0>

```

Not sure how it happened though...

MichaelBitard commented on nvim-plugin/lua/ethersync/cursor.lua:
```diff
@@ -233,6 +247,79 @@ function M.track_cursor(bufnr, callback)
+    following_userid = nil
+end
+
+function M.follow_cursor()
```
Yeah, I saw that and was afraid on how callback worked in lua. But it was quite simple

blinry replied:
I like how it turned out! :)

MichaelBitard commented on nvim-plugin/lua/ethersync/cursor.lua:
```diff
@@ -233,6 +247,79 @@ function M.track_cursor(bufnr, callback)
+                local line_number = vim.fn.line(".")
+                local location = locations[line_number]
+                vim.api.nvim_win_close(win, true)
+                jump(location)
```
Yeap, that was a mistake

blinry commented on nvim-plugin/README.md:
```diff
@@ -31,9 +31,10 @@ Usually, you will add the string `"ethersync/ethersync-nvim"` to your plugin man
-    { "<leader>u", "<cmd>EthersyncStopFollow<cr>" },
+    { "<leader>ej", "<cmd>EthersyncJumpToCursor<cr>" },
+    { "<leader>ef", "<cmd>EthersyncFollow<cr>" },
+    { "<leader>eu", "<cmd>EthersyncUnfollow<cr>" },
```
My feeling is: If we have the <esc> mapping, we don't even need to suggest the <leader>eu one. Less mappings to keep track of.

blinry reviewed (approved):
Works perfectly for me! :) Thank you! I have another suggestion regarding the example mappings, but that's kind of optional.

To wrap this up, could you squash your commits together? (If that's any trouble, I can also do it when merging.)

And feel free to add yourself to the copyright header of cursor.lua – credit where credit is due! :) (We're following [the REUSE spec](https://reuse.software/)!)

MichaelBitard commented:
Thanks for everything!

I have squashed the commits, added myself to the copyrights and removed the `<leader>eu` keymap!

I hope we'll keep using it and that I'll be able to contribute again on this, it's been a pleasure!
